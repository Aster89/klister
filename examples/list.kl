#lang "n-ary-app.kl"

(import "defun.kl")


(define car
  (lambda (stx)
    (syntax-case stx
      ((cons a _)
       a))))

(define cdr
  (lambda (stx)
    (syntax-case stx
      ((cons _ d)
       d))))

(defun map (f stx)
  (syntax-case stx
    (() stx)
    ((cons a d)
     (cons-list-syntax (f a) (map f d) stx))))

(example (car '(1 2 3)))
(example (cdr '(1 2 3)))
(example (map car '((x 1) (y 2) (z 3))))

(export car cdr map)
