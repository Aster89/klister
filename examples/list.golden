'Signal 1 : Syntax
'(Signal 2 Signal 3) : Syntax
'(x y z) : Syntax
'((f x) (f y) (f z)) : Syntax
'(z y x) : Syntax
